CLI implementation for AutoDR (Failover Groups)

maboja-msft opened the pull request:
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

promptws commented:
View a preview at https://prompt.ws/r/Azure/azure-cli/5710
This is an **experimental preview** for @microsoft users.

maboja-msft commented:
(env) C:\azure-cli>az sql failover-group -h

Group
    az sql failover-group: Manage SQL Failover Groups.

Commands:
    add-databases   : Adds provided list of databases to the failover group.
    create          : Creates a failover group.
    delete          : Deletes a failover group.
    failover        : Executes a failover of a failover group.
    list            : Lists the failover groups in a server.
    remove-databases: Removes provided list of databases from failover group.
    show            : Gets a failover group.
    update          : Updates the parameters of the failover group.


(env) C:\azure-cli>az sql failover-group create -h

Command
    az sql failover-group create: Creates a failover group.

Arguments
    --name -n                  [Required]: The name of the Failover Group.
    --partner-server-name      [Required]: The name of the partner server of a Failover Group.
    --resource-group -g        [Required]: Name of resource group. You can configure the default
                                           group using `az configure --defaults group=<name>`.
    --server -s                [Required]: Name of the Azure SQL server.
    --allow-read-only-failover-to-primary: This value indicates whether an outage on the secondary
                                           server should trigger automatic failover of the read-only
                                           endpoint. This feature is not yet supported.
    --failover-policy                    : The failover policy of the Failover Group.  Default:
                                           Automatic.
    --grace-period-with-data-loss        : Interval before automatic failover is initiated if an
                                           outage occurs on the primary server. This indicates that
                                           Azure SQL Database will not initiate automatic failover
                                           before the grace period expires. Please note that
                                           failover operation with AllowDataLoss option might cause
                                           data loss due to the nature of asynchronous
                                           synchronization.  Default: 60.
    --partner-resource-group-name        : The name of the resource group of the partner server.

Global Arguments
    --debug                              : Increase logging verbosity to show all debug logs.
    --help -h                            : Show this help message and exit.
    --output -o                          : Output format.  Allowed values: json, jsonc, table, tsv.
                                           Default: json.
    --query                              : JMESPath query string. See http://jmespath.org/ for more
                                           information and examples.
    --verbose                            : Increase logging verbosity. Use --debug for full debug
                                           logs.

(env) C:\azure-cli>az sql failover-group update -h

Command
    az sql failover-group update: Updates the parameters of the failover group.

Arguments
    --name -n                  [Required]: The name of the Failover Group.
    --resource-group -g        [Required]: Name of resource group. You can configure the default
                                           group using `az configure --defaults group=<name>`.
    --server -s                [Required]: Name of the Azure SQL server.
    --allow-read-only-failover-to-primary: This value indicates whether an outage on the secondary
                                           server should trigger automatic failover of the read-only
                                           endpoint. This feature is not yet supported.
    --failover-policy                    : The failover policy of the Failover Group.
    --grace-period-with-data-loss        : Interval before automatic failover is initiated if an
                                           outage occurs on the primary server. This indicates that
                                           Azure SQL Database will not initiate automatic failover
                                           before the grace period expires. Please note that
                                           failover operation with AllowDataLoss option might cause
                                           data loss due to the nature of asynchronous
                                           synchronization.

Global Arguments
    --debug                              : Increase logging verbosity to show all debug logs.
    --help -h                            : Show this help message and exit.
    --output -o                          : Output format.  Allowed values: json, jsonc, table, tsv.
                                           Default: json.
    --query                              : JMESPath query string. See http://jmespath.org/ for more
                                           information and examples.
    --verbose                            : Increase logging verbosity. Use --debug for full debug
                                           logs.

(env) C:\azure-cli>az sql failover-group add-databases -h

Command
    az sql failover-group add-databases: Adds provided list of databases to the failover group.

Arguments
    --databases         [Required]: List of databases to add (for add-databases command)or remove
                                    (for remove-databases command) from Failover Group.
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) C:\azure-cli>az sql failover-group remove-databases -h

Command
    az sql failover-group remove-databases: Removes provided list of databases from failover group.

Arguments
    --databases         [Required]: List of databases to add (for add-databases command)or remove
                                    (for remove-databases command) from Failover Group.
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) C:\azure-cli>az sql failover-group failover -h

Command
    az sql failover-group failover: Executes a failover of a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --allow-data-loss             : Complete the failover even if doing so may result in data loss.
                                    This will allow the failover to proceed even if a primary
                                    database is unavailable.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) C:\azure-cli>az sql failover-group delete -h

Command
    az sql failover-group delete: Deletes a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) C:\azure-cli>az sql failover-group show -h

Command
    az sql failover-group show: Gets a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) C:\azure-cli>az sql failover-group list -h

Command
    az sql failover-group list: Lists the failover groups in a server.

Arguments
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

maboja-msft commented:
@jaredmoo  @btasdoven  Can you please take a look?

jaredmoo commented:
I want to make sure that the design works nicely with future managed failover group and job target group commands. Specifically

- What will be the planned syntax for managed failover group? (Managed failover group will be able to contain both MI and servers/db, right?)
- `failover-group add-databases --databases x y z` a few comments:
  - I think adding/removing should be part of `failover-group update`. e.g. `failover-group update --add-db  x y z`. This allows multiple properties to be set in one go. 
  - `failover-group create` also should have the ability to add db's
- `--partner-server-name` -> `--partner-server`
- `--partner-resource-group-name` -> `--partner-resource-group`
- Please comment what happens if `--partner-resource-group` is not specified
- `--allow-read-only-failover-to-primary: (...) This feature is not yet supported.` - so why is there a param for it? :)
- `--failover-policy` - what are allowed values? Can you put `enum` in params.py?
- `failover`
  - I know "failover" is in the name of the resource, but please consider `set-primary` in order to match same as `az sql db replica set-primary` command (single db geodr). 
  - please make sure it is idempotent.
- `--grace-period-with-data-loss` this is very long param name. We are not implementing powershell here ;) can it just be `--grace-period` ?

maboja-msft commented:
@jaredmoo I have addressed issues you mentioned in your comment. I don't know however how CLI for managed instances would look like and the team working on it does not know either. So imho we shouldn't take a dependency on it.

jaredmoo commented:
Managed failover group is not that far away, it would be bad to add failover group now and then realise that managed failover group needs to be totally different. Can you past help output so I can review updated design?

maboja-msft commented:
I talked with manged failover group folks and they are ok with following the design for sterling. 
@btasdoven @dlemMSFT Can you look at this PR from GeoDR team point of view?

Here is updated help output (without global arguments which are staying default):
(env) C:\azure-cli>az sql failover-group -h

Group
    az sql failover-group: Manage SQL Failover Groups.

Commands:
    create     : Creates a failover group.
    delete     : Deletes a failover group.
    list       : Lists the failover groups in a server.
    set-primary: Executes a failover of a failover group.
    show       : Gets a failover group.
    update     : Updates the parameters of the failover group.

(env) C:\azure-cli>az sql failover-group create -h

Command
    az sql failover-group create: Creates a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --partner-server    [Required]: The name of the partner server of a Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --add-db                      : List of databases to add to Failover Group.
    --failover-policy             : The failover policy of the Failover Group.  Allowed values:
                                    Automatic, Manual.  Default: Automatic.
    --grace-period                : Interval before automatic failover is initiated if an outage
                                    occurs on the primary server. This indicates that Azure SQL
                                    Database will not initiate automatic failover before the grace
                                    period expires. Please note that failover operation with
                                    AllowDataLoss option might cause data loss due to the nature of
                                    asynchronous synchronization.  Default: 60.
    --partner-resource-group      : The name of the resource group of the partner server.


(env) C:\azure-cli>az sql failover-group update -h

Command
    az sql failover-group update: Updates the parameters of the failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --add-db                      : List of databases to add to Failover Group.
    --failover-policy             : The failover policy of the Failover Group.  Allowed values:
                                    Automatic, Manual.
    --grace-period                : Interval before automatic failover is initiated if an outage
                                    occurs on the primary server. This indicates that Azure SQL
                                    Database will not initiate automatic failover before the grace
                                    period expires. Please note that failover operation with
                                    AllowDataLoss option might cause data loss due to the nature of
                                    asynchronous synchronization.
    --remove-db                   : List of databases to remove from Failover Group.


(env) C:\azure-cli>az sql failover-group set-primary -h

Command
    az sql failover-group set-primary: Executes a failover of a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --allow-data-loss


(env) C:\azure-cli>az sql failover-group delete -h

Command
    az sql failover-group delete: Deletes a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.


(env) C:\azure-cli>az sql failover-group show -h

Command
    az sql failover-group show: Gets a failover group.

Arguments
    --name -n           [Required]: The name of the Failover Group.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

btasdoven commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_help.py:
```diff
@@ -287,6 +287,22 @@
+"""
+helps['sql failover-group set-primary'] = """
+    type: command
+    short-summary: Executes a failover of a failover group.
```
Can we make this more clear and similar to set-primary parameter in GeoDr links?
`Set the primary replica database by failing over from the current primary replica database.`

jaredmoo commented:
Design looks good!
- --grace-period: please comment in help that it is in seconds
- --allow-data-loss: missing help text

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_help.py:
```diff
@@ -287,6 +287,22 @@
+"""
+helps['sql failover-group update'] = """
+    type: command
+    short-summary: Updates the parameters of the failover group.
```
Updates a failover group

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py:
```diff
@@ -594,9 +595,31 @@ def _configure_security_policy_storage_params(arg_ctx):
         c.argument('database_dtu_max', help='The maximum DTU any one database can consume.')
         c.argument('database_dtu_min', help='The minimum DTU all databases are guaranteed.')
-        c.argument('dtu', help='TThe total shared DTU for the elastic eool.')
+        c.argument('dtu', help='The total shared DTU for the elastic eool.')
```
TThanks ;)

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py:
```diff
@@ -161,6 +162,20 @@ def load_command_table(self, _):
+        g.command('show', 'get')
+        g.command('list', 'list_by_server')
+        g.custom_command('create', 'failover_group_create')
+        g.custom_command('update', 'failover_group_update')
```
Why custom_command instead of generic_update_command?

jaredmoo replied:
Note that custom function for generic_updated_command does not work the same as custom_command

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -1010,6 +1014,165 @@ def elastic_pool_list_capabilities(
+        quote(server_name),
+        quote(d)) for d in remove_db]
+
+    databases = list((set(databases_in_fg) | set(add_db_ids)) - set(remove_db_ids))
```
Make sure this is case insensitive

derekbekoe commented:
This PR looks stale. Closing soon...
